src/computedevs.jl

```diff
@@ -0,0 +1,80 @@
+# adaptors for converting abstract arrays to Base.Array
+
+
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
+
+`adapt(dev::AbstractComputingDevice, x)` adapts `x` for `dev`.
+"""
+abstract type AbstractComputingDevice end
+export AbstractComputingDevice
+
+
+"""
+    struct ComputingDeviceIndependent <: AbstractComputingDevice
+
+`get_computing_device(x) === ComputingDeviceIndependent()` indicates
+that `x` is not tied to a specific computing device. This typically
+means that x is a statically allocated object.
+"""
+struct ComputingDeviceIndependent <: AbstractComputingDevice end
+export ComputingDeviceIndependent
+
+
+"""
+    struct CPUDevice <: AbstractComputingDevice
+
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
+
+
+
+"""
+    abstract type AbstractGPUDevice <: AbstractComputingDevice
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputingDevice end
+export AbstractGPUDevice
+
+
+
+const _incompatible_devs = ArgumentError("Incompatible computing devices")
+
+select_computing_device(a::ComputingDeviceIndependent, ::ComputingDeviceIndependent) = a
+select_computing_device(a::ComputingDeviceIndependent, b::AbstractComputingDevice) = b
+select_computing_device(a::AbstractComputingDevice, b::ComputingDeviceIndependent) = a
+
+select_computing_device(a::CPUDevice, ::CPUDevice) = a
+select_computing_device(a::CPUDevice, b::AbstractGPUDevice) = a
+select_computing_device(a::AbstractGPUDevice, b::CPUDevice) = b
+select_computing_device(a::AbstractGPUDevice, b::AbstractGPUDevice) = (a === b) ? a : throw(_incompatible_devs)
+
+
+"""
+    get_computing_device(x)::AbstractComputingDevice
+
+Get the computing device backing object `x`.
+"""
+function get_computing_device end
+export get_computing_device
+
+
+@inline get_computing_device(::Array) = CPUDevice()
+
+# ToDo: Utilize `ArrayInterfaceCore.buffer(A)`? Would require Adapt to depend
+# on ArrayInterfaceCore.
+
+@generated function get_computing_device(x)
+    impl = :(begin dev_0 = ComputingDeviceIndependent() end)
+    append!(impl.args, [:($(Symbol(:dev_, i)) = select_computing_device(get_computing_device(getfield(x, $i)), $(Symbol(:dev_, i-1)))) for i in 1:fieldcount(x)])
+    push!(impl.args, :(return $(Symbol(:dev_, fieldcount(x)))))
+    impl
+end
```

Comment:
I would avoid this if I were doing this because:

* It doesn't handle cycles. (Cycles can be created for mutable structs. Recursing only for immutable is a bad idea since mutable vs immutable struct is an implementation detail.)
* There is nothing wrong with a given collection object containing references to "remote" memories. For example, it's possible to have a "CPU array" that is backed up by `Base.Array` but refers to a `CuArray` from which data is copied upon certain events. The data access happens purely through `Base.Array`.

If I were doing this, I would define a fold over "relevant data source objects" in a manner generalizing `Adapt.adapt_structure`.

Reply:
Good point about the loops. I'm also not perfectly happy with this magic, but on the other hand it's unreasonable to expect struct maintainers to implement a large number of methods to make their structures GPU+ML capable. They already have to implement `Adapt.adapt_storage`, Functors.jl and possibly ConstructionBase.jl support, etc ... that's another consensus API we're currently lacking in my opinion, semantic struct tree introspection, de- and re-composition.

> If I were doing this, I would define a fold over "relevant data source objects" in a manner generalizing Adapt.adapt_structure.

I guess there we enter Functors & friends territory ... it might be good is some key people from the array ecosystem would join the discussion in JuliaObjects/ConstructionBase.jl#53 (relevant parts at the end). I feel there's a lot of unused synergy potential and several parallel incompatible solutions in this area currently.

For now, if we want to avoid generated function magic in the PR, we could use `ArrayInterfaceCore.buffer` (which defaults to `Base.parent`). ArrayInterfaceCore.jl is now very lightweight, so I think it could be an acceptable dependency for Adapt.jl (CC @ChrisRackauckas ). So we could do something like

```julia
function get_computing_device(x::AbstractArray)
    buf = buffer(x)
    buf === x && throw(ArgumentError("Can't get buffer of object"))
    get_computing_device(buf)
end
```

Reply:
Or, with a bottom value instead of an exception (see suggestion of @tkf below):

```julia
function get_computing_device(x::AbstractArray)
    buf = ArrayInterfaceCore.buffer(x)
    if typeof(buf) != typeof(x)
        get_computing_device(buf)
    else
        UnknownComputingDevice(x)
    end
end
```

Reply:
@tkf I prototyped such a generic fold at https://github.com/FluxML/Functors.jl/pull/32 (and previously https://github.com/Flfolding and uxML/Functors.jl/pull/27). Part of the tension I ran into is that while we may see flattening as a special case of folding/`fmap`ping, the existing Julia ecosystem tends to see it the other way around (traversal happens after flattening).

For this particular PR, I agree the bottom value is the safest option. After all, what would `get_computing_device` return for a NN with mixed CPU and GPU array parameters otherwise?

Reply:
> After all, what would get_computing_device return for a NN with mixed CPU and GPU array parameters otherwise?

I think it should return "incompatible devices"/"can't find common device".

Reply:
> It doesn't handle cycles. 

Cycles are taken care of now.

> There is nothing wrong with a given collection object containing references to "remote" memories. For example, it's possible to have a "CPU array" that is backed up by Base.Array but refers to a CuArray from which data is copied upon certain events. [...]  I would define a fold over "relevant data source objects"

I would say such special types would specialize the new `get_compute_unit_impl`. A generic mechanism to get the "relevant data source objects" of an objects would be nice, of course, but we don't have an ecosystem-wide solution yet I think? When we get one though, we can just use it in the generic `get_compute_unit_impl`.

Reply:
I'll mark this resolved, since I've changed a lot of things, but please feel free to reopen.